ip_country/src/ip_country.rs

```diff
@@ -116,9 +119,151 @@ fn write_value(
+            self.inner_arc.clone()
+        }
 
-    static TEST_DATA: &str = "0.0.0.0,0.255.255.255,ZZ
+        pub fn get_bytes(&self) -> Vec<u8> {
+            self.inner_arc.lock().unwrap().byte_array.clone()
+        }
+
+        pub fn reject_next_write(&mut self, error: Error) {
+            self.inner_arc().lock().unwrap().next_error = Some(error);
+        }
+    }
+
+    impl Write for ByteArrayWriter {
+        fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+            let mut inner = self.inner_arc.lock().unwrap();
+            if let Some(next_error) = inner.next_error.take() {
+                Err(next_error)
+            } else {
+                for byte in buf {
+                    inner.byte_array.push(*byte)
+                }
+                Ok(buf.len())
+            }
+        }
+
+        fn flush(&mut self) -> io::Result<()> {
+            Ok(())
+        }
+    }
+
+    pub struct ByteArrayReader {
+        byte_array: Vec<u8>,
+        position: usize,
+        next_error: Option<Error>,
+    }
+
+    impl ByteArrayReader {
+        pub fn new(byte_array: &[u8]) -> ByteArrayReader {
+            ByteArrayReader {
+                byte_array: byte_array.to_vec(),
+                position: 0,
+                next_error: None,
+            }
+        }
+    }
+
+    impl Read for ByteArrayReader {
+        fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+            match self.next_error.take() {
+                Some(error) => Err(error),
+                None => {
+                    let to_copy = min(buf.len(), self.byte_array.len() - self.position);
+                    #[allow(clippy::needless_range_loop)]
+                    for idx in 0..to_copy {
+                        buf[idx] = self.byte_array[self.position + idx]
+                    }
+                    self.position += to_copy;
+                    Ok(to_copy)
+                }
+            }
+        }
+    }
+
+    impl BufRead for ByteArrayReader {
+        fn fill_buf(&mut self) -> io::Result<&[u8]> {
+            match self.next_error.take() {
+                Some(error) => Err(error),
+                None => Ok(&self.byte_array[self.position..]),
+            }
+        }
+
+        fn consume(&mut self, amt: usize) {
+            let result = self.position + amt;
+            self.position = if result < self.byte_array.len() {
+                result
+            } else {
+                self.byte_array.len()
+            }
+        }
+    }
+
+    pub struct FakeStreamHolder {
+        pub stdin: ByteArrayReader,
+        pub stdout: ByteArrayWriter,
+        pub stderr: ByteArrayWriter,
+    }
+
+    impl Default for FakeStreamHolder {
+        fn default() -> Self {
+            FakeStreamHolder {
+                stdin: ByteArrayReader::new(&[0; 0]),
+                stdout: ByteArrayWriter::new(),
+                stderr: ByteArrayWriter::new(),
+            }
+        }
+    }
+
+    impl FakeStreamHolder {}
+
```

Comment:
Everything added from here up to the beginning of the test module is wrong. Let's go with what I suggested in the parallel branch, when you pick up on my commit ending that branch this all will have been take care of. We do need to prevent this hacky way of fixes. That's all right, I know, you didn't mean it badly. I'm glad though we can go on without duplicating the code.